Remove unnecessary helper function that adds a function call to a code block

antlechner opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->
No functional change, only refactoring.
~The helper function is not directly related to nondet object creation and can be used in another function outside of the object factory.~ The function has now been removed entirely as it can be simplified and inlined.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- n/a My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 54d5bf448bdf0223a9e268e9c629f6f1386cd646).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105908403
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

owen-mc-diffblue commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -262,6 +262,15 @@ void add_method_call_if_present(
+  const symbol_table_baset &symbol_table)
+{
+  add_method_call_if_present(
+    assignments, nil_exprt{}, method_name, symbol_table);
```
:pick: I think it would be slightly nicer to use an optional `exprt` for the second argument, so we don't have to pass a `nil_exprt` like this. There could be a precondition that if the second argument is non-empty then the type has a `this` parameter.

antlechner replied:
Passing in `null` as an instance to call a method on if the method is static is quite common in Java (e.g. `java.lang.reflect`) but I agree that an optionalt is probably clearer in the context of JBMC. So I changed it to that. I added an invariant that's the converse of your suggestion: if the type has a `this` parameter (i.e. it represents an instance method), the second argument has to be non-empty. It still seems safe to just ignore the second argument if the represented Java method is static. But if you think it's clearer I could change it to something like
```
    if(instance_expr)
    {
      INVARIANT(
        type.has_this(),
        "A call to a code_function_callt representing a static Java method "
        "should not be given an instance argument.");
      fun_call.arguments().push_back(address_of_exprt(*instance_expr));
    }
    else
    {
      INVARIANT(
        !type.has_this(),
        "A call to a code_function_callt representing a Java instance method "
        "should be given an instance argument.");
    }
```
Let me know what you think.

antlechner replied:
Or I could just get rid of `instance_expr` altogether and make it part of the vector that @smowton suggested below? I'll try that.

smowton commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -246,6 +246,31 @@ void java_add_components_to_class(
     components.end(), components_to_add.begin(), components_to_add.end());
 }
 
+void add_method_call_if_present(
```
It's a bit weird to have a general-purpose utility function that only works on no-arg functions. I think you should take a vector-of-exprt and append those if given.

antlechner committed:
Inline gen_method_call_... in Java object factory

The body of this function can be simplified to use the constructor of
code_function_callt that takes a vector of arguments, at which point the
function becomes so simple that it makes sense to just inline it.

antlechner committed:
Simplify function calls in java_static_initializers

This can be written in fewer lines using symbol_table.lookup. The logic
is the same as in the object factory in the previous commit.

antlechner commented:
@owen-jones-diffblue @smowton As I tried to implement @smowton 's suggestion to add a vector for the arguments, and to make the `instance_exprt` the first element of that vector, the helper function actually became so simple that it made more sense to just inline it. So I did that and the PR is completely different now - please re-review. :slightly_smiling_face:

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 25abc15ee3530b836726b5df9a129298943fa718).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105996838
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 1c638651c1b2e4752c16e71ee4dfdcc4a5bbb9ed).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106006577
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
